certbot-ci/certbot_integration_tests/utils/acme.py

```diff
@@ -0,0 +1,215 @@
+"""Module to setup an ACME CA server environment able to run multiple tests in parallel"""
+from __future__ import print_function
+import tempfile
+import atexit
+import os
+import subprocess
+import shutil
+import sys
+from os.path import join
+
+import requests
+import json
+
+from certbot_integration_tests.utils import misc
+
+ACME_V1_PORT = 4000   # Implicitly configured in boulder, not used in pebble
+ACME_V2_PORT = 4001  # Implicitly configured in boulder, explicitly in pebble
+ACME_SUBNET = '10.77.77'  # Implicitly configured in boulder, explicitly in pebble
+CHALLTESTSRV_PORT = 8055  # Implicitly configured in boulder, explicitly in pebble
+HTTP_01_PORT = 5002  # Implicitly configured both in boulder and pebble
+
+
+def setup_acme_server(acme_config, nodes):
+    """
+    This method will setup an ACME CA server and a HTTP reverse proxy instances, to allow parallel
+    execution of integration tests against the unique http-01 port expected by the ACME CA server.
+    Instances are properly closed and cleaned when the Python process exits using atexit.
+    Typically all pytest integration tests will be executed in this context.
+    This method returns an object describing ports and directory url to use for each pytest node
+    with the relevant pytest xdist node.
+    :param dict acme_config: a dict describing the current acme server characteristics to setup.
+    :param str[] nodes: list of nodes name that will be setup by pytest xdist
+    :return: a dict describing the challenges ports that have been setup for the nodes
+    :rtype: dict
+    """
+    acme_type, acme_option = acme_config['type'], acme_config['option']
+    acme_xdist = _construct_acme_xdist(acme_type, acme_option, nodes)
+    workspace = _construct_workspace(acme_type)
+
+    _prepare_traefik_proxy(workspace, acme_xdist)
+    _prepare_acme_server(workspace, acme_type, acme_option, acme_xdist)
+
+    return acme_xdist
+
+
+def _construct_acme_xdist(acme_type, acme_option, nodes):
+    """Generate and return the acme_xdist dict"""
+    acme_xdist = {'challtestsrv_port': CHALLTESTSRV_PORT}
+
+    if acme_type == 'pebble':
+        acme_xdist['directory_url'] = 'https://localhost:{0}/dir'.format(ACME_V2_PORT)
+    else:  # boulder
+        port = ACME_V2_PORT if acme_option == 'v2' else ACME_V1_PORT
+        acme_xdist['directory_url'] = 'http://localhost:{0}/directory'.format(port)
+
+    acme_xdist['http_port'] = {node: port for (node, port)
+                               in zip(nodes, range(5200, 5200 + len(nodes)))}
```

Comment:
why 5100 and 5200? and I assume we're sure we'll have fewer than 100 nodes? :)

Reply:
Because I like this symetry ^^
5001 => 5100 - 5199
5002 => 5200 - 5299

Indeed, "only" 100 slots each, I dare you to find a machine that has enough power to handle that :p